File: scripts/route.py
#!/usr/bin/env python3
"""Auto-router: measure the diff, then pick the brief or the full map.

  small change  -> impact_brief.py  (read the diff yourself; get the cross-repo
                   signals — blast radius, breaking changes, test gap).
  large change  -> full map         (diff_summary + symbol_impact, with the
                   impact-aware review_order to triage what to read).

The threshold is a heuristic from evals/FINDINGS.md: the full map only pays off
once the diff is too big to just read directly. Tune with --max-brief-files /
--max-brief-lines, or pin a path with --force {brief,full}.

Usage:
    python scripts/route.py --range main..HEAD --cwd <repo>
    python scripts/route.py --staged --cwd <repo> --json
    python scripts/route.py --range main..HEAD --cwd <repo> --force full

This is the recommended single entry point — call it first; it decides.
"""

import argparse
import json
import os
import subprocess
import sys
import tempfile

SCRIPTS = os.path.dirname(os.path.abspath(__file__))
DEFAULT_MAX_FILES = 3
DEFAULT_MAX_LINES = 60


def run(script, *args):
    cp = subprocess.run([sys.executable, os.path.join(SCRIPTS, script), *map(str, args)],
                        stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    return cp.returncode, cp.stdout


def run_json(script, *args):
    _, out = run(script, *args)
    try:
        return json.loads(out.strip())
    except ValueError:
        return None


def main():
    p = argparse.ArgumentParser(description="Auto-route a review to brief or full map")
    p.add_argument("--range", help="git range base..head")
    p.add_argument("--staged", action="store_true", help="route the staged index")
    p.add_argument("--cwd", default=".", help="repository working directory")
    p.add_argument("--json", action="store_true", help="emit JSON instead of text")
    p.add_argument("--max-brief-files", type=int, default=DEFAULT_MAX_FILES,
                   help="use the brief when <= this many files change (default 3)")
    p.add_argument("--max-brief-lines", type=int, default=DEFAULT_MAX_LINES,
                   help="use the brief when <= this many lines change (default 60)")
    p.add_argument("--force", choices=["brief", "full"], help="skip routing")
    args = p.parse_args()

    cwd = os.path.abspath(args.cwd)
    scope = ["--staged"] if args.staged else (["--range", args.range] if args.range else [])

    # Cheap measurement pass (Layer 1).
    l1 = run_json("diff_summary.py", "--cwd", cwd, "--compact", *scope)
    if l1 is None:
        sys.stderr.write("error: could not read the diff\n")
        sys.exit(2)
    totals = l1.get("totals", {})
    files = totals.get("files", 0)
    lines = totals.get("additions", 0) + totals.get("deletions", 0)

    small = files <= args.max_brief_files and lines <= args.max_brief_lines
    mode = args.force or ("brief" if small else "full")
    reason = ("forced" if args.force else
              "%d file(s), %d changed line(s); brief when <=%d files and <=%d lines"
              % (files, lines, args.max_brief_files, args.max_brief_lines))

    if mode == "brief":
        payload = run_json("impact_brief.py", "--cwd", cwd, "--json", *scope)
        if args.json:
            json.dump({"mode": "brief", "reason": reason, "brief": payload},
                      sys.stdout, indent=2)
            sys.stdout.write("\n")
        else:
            print("mode: brief (%s)" % reason)
            _, text = run("impact_brief.py", "--cwd", cwd, *scope)
            sys.stdout.write(text)
        return

    # full map: reuse the L1 we already computed; add Layer 2 impact.
    with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as fh:
        json.dump(l1, fh)
        l1_path = fh.name
    try:
        l2 = run_json("symbol_impact.py", "--root", cwd, "--diff-json", l1_path, "--compact")
    finally:
        os.unlink(l1_path)

    review_order = (l2 or {}).get("review_order") or l1.get("review_order", [])
    out = {
        "mode": "full",
        "reason": reason,
        "totals": totals,
        "risk_flags": l1.get("risk_flags", []),
        "review_order": review_order,
        "impact_flags": (l2 or {}).get("impact_flags", []),
        "prioritized": (l2 or {}).get("prioritized", []),
    }
    if args.json:
        json.dump(out, sys.stdout, indent=2)
        sys.stdout.write("\n")
        return

    print("mode: full map (%s)" % reason)
    if out["risk_flags"]:
        print("risk: " + ", ".join(out["risk_flags"]))
    if out["impact_flags"]:
        print("impact: " + ", ".join(out["impact_flags"]))
    print("review in this order (highest priority first):")
    for r in (out["prioritized"][:10] or [{"path": p} for p in review_order[:10]]):
        reasons = ("  — " + "; ".join(r["reasons"])) if r.get("reasons") else ""
        print("  %s%s" % (r["path"], reasons))
    print("\n(fetch hunks for the top files via diff_summary; this is the triage order)")


if __name__ == "__main__":
    main()

File: js/route.mjs
#!/usr/bin/env node
// Auto-router (recommended entry point) — Node port of scripts/route.py.
// Measures the diff and its blast radius, then picks brief vs full map; large
// changesets get the pattern-compression + checks + fan-out toolkit.

import { execFileSync } from "node:child_process";
import { writeFileSync, unlinkSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { pathToFileURL, fileURLToPath } from "node:url";

const JS = path.dirname(fileURLToPath(import.meta.url));
const DEFAULT_MAX_FILES = 3, DEFAULT_MAX_LINES = 60, DEFAULT_MIN_BLAST = 10, DEFAULT_LARGE_FILES = 25;
const FANOUT_BATCH = 10;

function run(script, args) {
  try { return execFileSync("node", [path.join(JS, script), ...args], { encoding: "utf8", maxBuffer: 1 << 28 }); }
  catch (e) { return e.stdout || ""; }
}
function runJson(script, args) {
  const out = run(script, args);
  try { return JSON.parse(out.trim()); } catch { return null; }
}

function buildLargeScale(cwd, scope, l1, l2, hasRange) {
  const patterns = runJson("diff_patterns.mjs", ["--cwd", cwd, "--json", ...scope]) || {};
  const patList = patterns.patterns || [];
  const unique = patterns.unique || [];
  const units = [...patList.map((p) => p.example_file), ...unique];
  const batches = [];
  for (let i = 0; i < units.length; i += FANOUT_BATCH) batches.push(units.slice(i, i + FANOUT_BATCH));

  let breaking = [];
  if (hasRange) {
    const rc = runJson("refactor_check.mjs", ["--range", scope[1], "--cwd", cwd]);
    if (rc && rc.ok) {
      const inv = rc.invariants || {};
      breaking = [...new Set([...(inv.public_signatures_changed || []), ...(inv.public_api_removed || [])])].sort();
    }
  }

  let highImpact = [];
  if (l2 && l2.ok) {
    const pub = (l2.symbols || []).filter((s) => s.exported && s.blast_radius > 0).sort((a, b) => b.blast_radius - a.blast_radius);
    highImpact = pub.slice(0, 20).map((s) => ({ symbol: s.name, file: s.file, blast_radius: s.blast_radius }));
  }

  return {
    compression: patterns.compression || {},
    patterns: patList.map((p) => ({ count: p.count, example_file: p.example_file, files: p.files, example_hunk: p.example_hunk })),
    unique, high_impact: highImpact, review_batches: batches,
    checks: { risk_flags: l1.risk_flags || [], impact_flags: (l2 || {}).impact_flags || [], breaking_changes: breaking },
  };
}

function parseArgs(argv) {
  const a = { range: null, staged: false, cwd: ".", json: false,
    maxBriefFiles: DEFAULT_MAX_FILES, maxBriefLines: DEFAULT_MAX_LINES,
    minBlast: DEFAULT_MIN_BLAST, largeFiles: DEFAULT_LARGE_FILES, force: null };
  for (let i = 0; i < argv.length; i++) {
    const x = argv[i];
    if (x === "--range") a.range = argv[++i];
    else if (x === "--staged") a.staged = true;
    else if (x === "--cwd") a.cwd = argv[++i];
    else if (x === "--json") a.json = true;
    else if (x === "--max-brief-files") a.maxBriefFiles = +argv[++i];
    else if (x === "--max-brief-lines") a.maxBriefLines = +argv[++i];
    else if (x === "--min-blast") a.minBlast = +argv[++i];
    else if (x === "--large-files") a.largeFiles = +argv[++i];
    else if (x === "--force") a.force = argv[++i];
  }
  return a;
}

function main() {
  const a = parseArgs(process.argv.slice(2));
  const cwd = path.resolve(a.cwd);
  const scope = a.staged ? ["--staged"] : (a.range ? ["--range", a.range] : []);

  const l1 = runJson("diff_summary.mjs", ["--cwd", cwd, "--compact", ...scope]);
  if (!l1) { process.stderr.write("error: could not read the diff\n"); process.exit(2); }
  const totals = l1.totals || {};
  const files = totals.files || 0;
  const lines = (totals.additions || 0) + (totals.deletions || 0);

  let l2cache;
  const getL2 = () => {
    if (l2cache === undefined) {
      const l1Path = path.join(tmpdir(), `crd-route-${process.pid}-${Date.now()}.json`);
      writeFileSync(l1Path, JSON.stringify(l1));
      try { l2cache = runJson("symbol_impact.mjs", ["--root", cwd, "--diff-json", l1Path, "--compact"]); }
      finally { try { unlinkSync(l1Path); } catch { /* */ } }
    }
    return l2cache;
  };

  const small = files <= a.maxBriefFiles && lines <= a.maxBriefLines;
  let mode, reason;
  if (a.force) { mode = a.force; reason = "forced"; }
  else if (small) { mode = "brief"; reason = `${files} file(s), ${lines} line(s) <= brief thresholds (${a.maxBriefFiles}/${a.maxBriefLines})`; }
  else if (files >= a.largeFiles) { mode = "full"; reason = `${files} file(s) >= ${a.largeFiles} -> full (ordering helps)`; }
  else {
    const l2 = getL2();
    let maxBlast = 0;
    if (l2 && l2.ok) maxBlast = Math.max(0, ...(l2.symbols || []).filter((s) => s.exported).map((s) => s.blast_radius));
    mode = maxBlast >= a.minBlast ? "full" : "brief";
    reason = `${files} file(s), ${lines} line(s), max blast radius ${maxBlast} -> ${mode} (full when blast >= ${a.minBlast})`;
  }

  if (mode === "brief") {
    if (a.json) {
      const payload = runJson("impact_brief.mjs", ["--cwd", cwd, "--json", ...scope]);
      process.stdout.write(JSON.stringify({ mode: "brief", reason, brief: payload }, null, 2) + "\n");
    } else {
      console.log(`mode: brief (${reason})`);
      process.stdout.write(run("impact_brief.mjs", ["--cwd", cwd, ...scope]));
    }
    return;
  }

  const l2 = getL2();
  const reviewOrder = (l2 || {}).review_order || l1.review_order || [];
  const out = {
    mode: "full", reason, totals,
    risk_flags: l1.risk_flags || [], review_order: reviewOrder,
    impact_flags: (l2 || {}).impact_flags || [], prioritized: (l2 || {}).prioritized || [],
  };
  if (files >= a.largeFiles) {
    const hasRange = !!a.range && !a.staged;
    const ls = buildLargeScale(cwd, scope, l1, l2, hasRange);
    out.large_scale = ls;
    if ((ls.patterns && ls.patterns.length) || (ls.unique && ls.unique.length)) {
      delete out.prioritized; delete out.review_order;
    }
  }

  if (a.json) { process.stdout.write(JSON.stringify(out, null, 2) + "\n"); return; }

  console.log(`mode: full map (${reason})`);
  if (out.risk_flags.length) console.log("risk: " + out.risk_flags.join(", "));
  if (out.impact_flags.length) console.log("impact: " + out.impact_flags.join(", "));
  const ls = out.large_scale;
  if (ls) {
    const c = ls.compression;
    console.log("large-scale toolkit:");
    console.log(`  patterns: ${c.changed_files || 0} changed files -> ${c.distinct_units || 0} distinct review units (${c.collapsed_files || 0} collapsed)`);
    ls.patterns.slice(0, 5).forEach((pat, i) => console.log(`    pattern ${i + 1}: ${pat.count} files like ${pat.example_file}`));
    if (ls.checks.breaking_changes.length) console.log("  BREAKING: " + ls.checks.breaking_changes.join(", "));
    console.log(`  fan-out: review ${ls.review_batches.length} batch(es) of <=${FANOUT_BATCH} units in isolated context`);
    if (ls.high_impact.length) {
      console.log("  high impact:");
      for (const h of ls.high_impact.slice(0, 10)) console.log(`    ${h.symbol} (${h.file}) — ${h.blast_radius} caller file(s)`);
    }
    return;
  }
  console.log("review in this order (highest priority first):");
  const list = out.prioritized.length ? out.prioritized.slice(0, 10) : reviewOrder.slice(0, 10).map((p) => ({ path: p }));
  for (const r of list) {
    const reasons = r.reasons && r.reasons.length ? "  — " + r.reasons.join("; ") : "";
    console.log(`  ${r.path}${reasons}`);
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) main();
